### PR TITLE
Non_exhaustive for plurals

### PR DESCRIPTION
**Pull request by Manishearth**
Split out of #1789

Progress on #170, #383

### CODE REVIEW

File: components/plurals/src/operands.rs [hunk 1]
```diff
@@ -33,45 +33,46 @@ use fixed_decimal::FixedDecimal;
...
+///    2, // v
+///    2, // w
+///    45, // f
+///    45, // t
+///    0, // c
+/// )), FixedDecimal::from(12345).multiplied_pow10(-2).map(|d| (&d).into()))
 /// ```
-#[derive(Debug, Clone, Copy, PartialEq)]
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+#[non_exhaustive]
```
Comment (Manishearth) [inline components/plurals/src/operands.rs:75]: cc @sffc @zbraniecki for discussing this choice

(from the previous thread: shane feels that this should be non_exhaustive, zibi feels that it should not)

**Comment by Manishearth** [comment]
@zbraniecki review ping

File: components/plurals/src/lib.rs [hunk 2]
```diff
@@ -95,6 +97,7 @@ use rules::runtime::test_rule;
 /// A type of a plural rule which can be associated with the [`PluralRules`] struct.
 ///
 #[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
+#[allow(clippy::exhaustive_enums)] // this type is stable
```
Comment (sffc) [inline components/plurals/src/lib.rs:100]: Nit: This looks like the kind of thing that we may want to extend in the future, if CLDR adds more types of plural forms.

File: components/plurals/src/lib.rs [hunk 3]
```diff
@@ -147,6 +150,7 @@ pub enum PluralRuleType {
 #[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, Ord, PartialOrd)]
 #[cfg_attr(feature = "datagen", derive(serde::Serialize))]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[allow(clippy::exhaustive_enums)] // this type is stable
```
Comment (sffc) [inline components/plurals/src/lib.rs:153]: Nit: Please explain a bit more why we are treating this as stable. @zbraniecki

File: components/plurals/src/operands.rs [hunk 4]
```diff
@@ -88,6 +89,10 @@ pub struct PluralOperands {
 }
 
 impl PluralOperands {
+    /// Construct a new [`PluralOperands`] given its various fields (documented on the struct)
+    pub fn new(i: u64, v: usize, w: usize, f: u64, t: u64, c: usize) -> Self {
```
Comment (sffc) [inline components/plurals/src/operands.rs:93]: Nit: This type of constructor, with a bunch of homogeneous arguments, is extremely messy and hard to read. I'd prefer if we kept `Construct` on `Default::default()` without the constructor. Alternatively, we can make the struct exhaustive.

File: components/plurals/src/operands.rs (hunk 1 above)
Comment (sffc) [reply components/plurals/src/operands.rs:75]: I'm torn on this. @eggrobin pointed out that if CLDR did decide to add an additional operand, we would have to figure out how to deal with data coming from CLDR with that operand, including how to gracefully fall back in patterns that contain the operand. Even if we held this back for an ICU4X major version bump, we still need to reason about generating data for older ICU4X versions for a certain period of time.

File: components/plurals/src/rules/runtime/ast.rs [hunk 5]
```diff
@@ -14,17 +14,20 @@ use zerovec::{
 };
 
 #[derive(yoke::Yokeable, zerofrom::ZeroFrom, Clone, PartialEq, Debug)]
+#[allow(clippy::exhaustive_structs)] // this type is stable
 pub struct Rule<'data>(pub VarZeroVec<'data, RelationULE>);
```
Comment (sffc) [inline components/plurals/src/rules/runtime/ast.rs:18]: Suggestion: The VarZeroVec should be a private field, and all the other types in this module should be pub(crate).

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by sffc — APPROVED** [review]

File: components/plurals/tests/plurals.rs [hunk 6]
```diff
@@ -20,19 +16,17 @@ fn test_plural_rules() {
...
+    let _rules: DataPayload<CardinalV1Marker> = provider
         .load_resource(&DataRequest {
             options: locale!("en").into(),
             metadata: Default::default(),
         })
         .expect("Failed to load payload")
         .take_payload()
         .expect("Failed to retrieve payload");
-    let rules = rules.get();
-    assert!(matches!(rules.one, Some(Rule(VarZeroVec::Borrowed(_)))));
```
Comment (sffc) [inline components/plurals/tests/plurals.rs:35]: Note to self: The borrowing behavior is tested elsewhere now.

File: components/plurals/tests/fixtures/mod.rs [hunk 7]
```diff
@@ -81,7 +81,7 @@ impl From<PluralOperandsInput> for PluralOperands {
                 f,
                 t,
                 c,
-            } => Self {
```
Comment (zbraniecki) [inline components/plurals/tests/fixtures/mod.rs:84]: `question`: why this change?

**Review by zbraniecki — APPROVED** [review]
lgtm, except of two changes from `Self` to named type that seem irrelevant (and, coincidentally, against my preference :))

File: components/plurals/tests/fixtures/mod.rs [hunk 8]
```diff
@@ -65,7 +65,7 @@ pub enum PluralOperandsInput {
 impl From<PluralOperandsInput> for PluralOperands {
     fn from(input: PluralOperandsInput) -> Self {
         match input {
-            PluralOperandsInput::List(operands) => Self {
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/plurals/tests/fixtures/mod.rs:68]: It's because the code got touched in intermediate commits that got undone.

Personally I prefer the explicit one for reading, though I don't like typing it so I also use Self quite a bit.
